gmock issue building realtime_tools on arch linux

I'm trying to build [ros-melodic-realtime-tools](https://aur.archlinux.org/packages/ros-melodic-realtime-tools/) from the arch linux AUR. [The PKGBUILD](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=ros-melodic-realtime-tools) shows how the make commands are called.

I get the following error:
```
$ makepkg
==> Making package: ros-melodic-realtime-tools 1.15.0-0 (Mon 09 Sep 2019 01:57:27 AM CDT)
==> Checking runtime dependencies...
==> Checking buildtime dependencies...
==> Retrieving sources...
  -> Found ros-melodic-realtime-tools-1.15.0.tar.gz
==> Validating source files with sha256sums...
    ros-melodic-realtime-tools-1.15.0.tar.gz ... Passed
==> Extracting sources...
  -> Extracting ros-melodic-realtime-tools-1.15.0.tar.gz with bsdtar
==> Removing existing $pkgdir/ directory...
==> Starting build()...
-- Using CATKIN_DEVEL_PREFIX: /home/jwhendy/installed/build/ros-melodic-aur/ros-melodic-realtime-tools/src/build/devel
-- Using CMAKE_PREFIX_PATH: /opt/ros/melodic
-- This workspace overlays: /opt/ros/melodic
-- Using PYTHON_EXECUTABLE: /usr/bin/python3
-- Using default Python package layout
-- Using empy: /usr/lib/python3.7/site-packages/em.py
-- Using CATKIN_ENABLE_TESTING: ON
-- Skip enable_testing() when building binary package
-- Using CATKIN_TEST_RESULTS_DIR: /home/jwhendy/installed/build/ros-melodic-aur/ros-melodic-realtime-tools/src/build/test_results
-- Found gtest: gtests will be built
-- Using Python nosetests: /usr/bin/nosetests-3.7
-- catkin 0.7.17
-- BUILD_SHARED_LIBS is on
CMake Warning at /opt/ros/melodic/share/catkin/cmake/test/gtest.cmake:151 (message):
  skipping gmock 'realtime_box_tests' in project 'realtime_tools' because
  gmock was not found
Call Stack (most recent call first):
  /opt/ros/melodic/share/catkin/cmake/test/gtest.cmake:80 (_catkin_add_executable_with_google_test)
  /opt/ros/melodic/share/catkin/cmake/test/gtest.cmake:54 (_catkin_add_google_test)
  CMakeLists.txt:23 (catkin_add_gmock)


CMake Error at CMakeLists.txt:24 (target_link_libraries):
  Cannot specify link libraries for target "realtime_box_tests" which is not
  built by this project.


-- Configuring incomplete, errors occurred!
See also "/home/jwhendy/installed/build/ros-melodic-aur/ros-melodic-realtime-tools/src/build/CMakeFiles/CMakeOutput.log".
See also "/home/jwhendy/installed/build/ros-melodic-aur/ros-melodic-realtime-tools/src/build/CMakeFiles/CMakeError.log".
```

I made a slight edit to a function to see where it was looking.
```
$ cat /opt/ros/melodic/share/catkin/cmake/test/gtest.cmake
function(_catkin_add_executable_with_google_test type target)
...
  if(NOT ${type_upper}_FOUND AND NOT ${type_upper}_FROM_SOURCE_FOUND)
    message(WARNING "skipping ${type} '${target}' in project '${PROJECT_NAME}' because ${type} was not found in '${GTEST_INCLUDE_DIRS}'")
```

This prints out as:
```
skipping gmock 'realtime_box_tests' in project 'realtime_tools' because
  gmock was not found in '/usr/include'
```

I do have gmock/gtest installed, and the dirs are in `/usr/include`. For the heck of it, I tried symlinking all `*.h` files into `/usr/include` directly, but that didn't work out.

```
$ pacman -Qi gmock
Name            : gmock
Version         : 1.8.1-3

$ pacman -Qi gtest
Name            : gtest
Version         : 1.8.1-3

$ ls /usr/include/gmock/
internal         gmock-cardinalities.h      gmock-generated-function-mockers.h  gmock-generated-nice-strict.h  gmock-matchers.h      gmock-more-matchers.h
gmock-actions.h  gmock-generated-actions.h  gmock-generated-matchers.h          gmock.h                        gmock-more-actions.h  gmock-spec-builders.h
```

I think this is related to #933 and #970. I also saw #1022  and adapted ros-melodic-catkin to build from @sloretz 's fork, but this did not solve my issue. I get the same original error about gmock not being found.

Any other details that would be helpful, or debug output I could add to assist with this?

I just re-built and installed `ros-melodic-catkin` looking at `kinetic-devel`, so I should be up to date:

```
### from pkg build directory
$ cat src/ros-melodic-catkin/CHANGELOG.rst | head

0.7.19 (2019-10-08)
-------------------
* support for pytest 5.1+ jUnit result files (`#1033 <https://github.com/ros/catkin/issues/1033>`_)
* fix -pthread handling in Debian buster (`#1021 <https://github.com/ros/catkin/issues/1021>`_)
* gmock from source on Debian buster (`#1022 <https://github.com/ros/catkin/issues/1022>`_)
* [windows] update a typo in local_setup.bat.in (`#1029 <https://github.com/ros/catkin/issues/1029>`_)
```

I get the same build error with `realtime_tools`:

```
CMake Warning at /opt/ros/melodic/share/catkin/cmake/test/gtest.cmake:151 (message):
  skipping gmock 'realtime_box_tests' in project 'realtime_tools' because
  gmock was not found
Call Stack (most recent call first):
  /opt/ros/melodic/share/catkin/cmake/test/gtest.cmake:80 (_catkin_add_executable_with_google_test)
  /opt/ros/melodic/share/catkin/cmake/test/gtest.cmake:54 (_catkin_add_google_test)
  CMakeLists.txt:23 (catkin_add_gmock)


CMake Error at CMakeLists.txt:24 (target_link_libraries):
  Cannot specify link libraries for target "realtime_box_tests" which is not
  built by this project.
```

And here is my attempt to understand why this fails.

[Here's where](https://github.com/ros/catkin/blob/77822d31cd84c052e776f58c4d5f5630e0760725/cmake/test/gtest.cmake#L268) it looks:
```
_catkin_find_google_source("gmock" "${_gmock_include_paths}"
                             "${_gmock_source_paths}" _gmock_found
                             _gmock_base_dir _gmock_include_dir _gmock_lib_dir
                             _gmock_libs _gmock_main_libs)
```
That [calls this](https://github.com/ros/catkin/blob/77822d31cd84c052e776f58c4d5f5630e0760725/cmake/test/gtest.cmake#L181):
```
function(_catkin_find_google_source type include_paths
  src_paths found base_dir include_dir lib_dir libs main_libs)
  # Find the gtest headers
  find_file(_${type}_INCLUDES "${type}.h"
            PATHS ${include_paths}
            NO_DEFAULT_PATH NO_CMAKE_FIND_ROOT_PATH)

  # Find the gtest sources
  find_file(_${type}_SOURCES "${type}.cc"
            PATHS ${src_paths}
            NO_DEFAULT_PATH NO_CMAKE_FIND_ROOT_PATH)
```

Thus, we look for two files from what I can tell:
```
find_file(_gmock__INCLUDES "gmock.h"
            PATHS ${include_paths}

find_file(_gmock__SOURCES "gmock.cc"
            PATHS ${src_paths}
```

At face value, do we have those files?
```
$ locate gmock.cc
/usr/src/gmock/gmock.cc

$ locate gmock.h
/usr/include/gmock/gmock.h
```

Ok, edit `gtest.cmake` to help us find out what it's looking for and where:
```
# Internal function for finding gtest or gmock sources
function(_catkin_find_google_source type include_paths
  src_paths found base_dir include_dir lib_dir libs main_libs)
  # Find the gtest headers
  message("##### _${type}_INCLUDES ${type}.h ${include_paths}")
  find_file(_${type}_INCLUDES "${type}.h"
            PATHS ${include_paths}
            NO_DEFAULT_PATH NO_CMAKE_FIND_ROOT_PATH)

  # Find the gtest sources
  message("##### _${type}_SOURCE ${type}.cc ${src_paths}")
  find_file(_${type}_SOURCES "${type}.cc"
            PATHS ${src_paths}
            NO_DEFAULT_PATH NO_CMAKE_FIND_ROOT_PATH)
```

What do we get?
```
##### _gtest_INCLUDES gtest.h /usr/include;/usr/src/gtest;/usr/include;/usr/src/googletest/googletest/include/gtest
##### _gtest_SOURCE gtest.cc /usr/src/gtest/src;/usr/src/googletest/googletest/src
##### _gmock_INCLUDES gmock.h /usr/include;/usr/src/gmock;/usr/include;/usr/src/googletest/googlemock/include/gmock
##### _gmock_SOURCE gmock.cc /usr/src/gmock/src;/usr/src/googletest/googlemock/src
```

So, it looks like it's looking or `/usr/include/{gtest,gmock}.h` and `/usr/src/{gtest, gmock}/src/{gtest, gmock}.cc` On arch linux, three were not in the right place according to these paths, but I fixed that with symlinks.
```
$ ls -l /usr/include/gtest.h 
lrwxrwxrwx 1 root root 26 Oct  8 21:30 /usr/include/gtest.h -> /usr/include/gtest/gtest.h
$ ls -l /usr/include/gmock.h
lrwxrwxrwx 1 root root 26 Oct  8 21:28 /usr/include/gmock.h -> /usr/include/gmock/gmock.h

$ ls -l /usr/src/gmock/src/gmock.cc 
lrwxrwxrwx 1 root root 23 Oct  8 21:34 /usr/src/gmock/src/gmock.cc -> /usr/src/gmock/gmock.cc

### gtest gets installed here by default, no need to symlink
$ ls -l /usr/src/gtest/src/gtest.cc 
-rw-r--r-- 1 root root 217839 Oct  1  2018 /usr/src/gtest/src/gtest.cc
```

Ah! Interesting. No error. I didn't expect that, and was pretty sure I did this up above. That said, printing out the includes from that function did leave me puzzled. Note [lines 258-259](https://github.com/ros/catkin/blob/77822d31cd84c052e776f58c4d5f5630e0760725/cmake/test/gtest.cmake#L258):
```
set(_gmock_include_paths "${include_path}/gmock")
set(_gmock_source_paths "${src_path}/gmock/src")
# ...
_catkin_find_google_source("gmock" "${_gmock_include_paths}"
                             "${_gmock_source_paths}" _gmock_found
                             _gmock_base_dir _gmock_include_dir _gmock_lib_dir
                             _gmock_libs _gmock_main_libs)
```

In my `message()` call above from within `function(_catkin_find_google_source type include_paths...)`, it looks to have been passed `include_paths=${_gmock_include_paths}`, which looks to include `${include_path}/gmock`. But only `/usr/include` was printed out. So, this issue looks like two issues:

- arch linux needs to add `/usr/src/gmock/src/{gmock-all.cc, gmock.cc, gmock_main.cc}` to their package.
- I _think_ `/usr/include/{gmock, gtest}/{gmock, gtest}.h` seems like a reasonable location for these files. What about a PR to add those to the include paths in `catkin`?
- I also think it's odd to have `/usr/src/gmock` in `_gmock_include_paths` (if it was in `_gmock_source_paths`, arch's default gmock location would work), and `/usr/include` is in there twice (should one have been `/usr/include/gmock`?). How about a bonus PR to clean that up, if possible?

> Maybe @sloretz can comment on that who authored that change.

It looks like he did, and then you merged it. I don't see any discussion of testing at all, nor does it look to refer to any issues so I'm not sure what specific issues prompted this.

I'm digging through internals more and am a bit out of my league to know what, exactly, needs fixing. I undid all my system changes, so I am back to the default file locations for [gmock](https://www.archlinux.org/packages/community/x86_64/gmock/) and [gtest](https://www.archlinux.org/packages/community/x86_64/gtest/) on arch linux.

My intuition was to change `_include_paths` to a list and append `/usr/src` vs. setting them at the same time. So, I started by changing [this line](https://github.com/ros/catkin/blob/kinetic-devel/cmake/test/gtest.cmake#L303) from this:
```
    set(_include_paths "/usr/include" "/usr/src")
```
... to this:
```
    set(_include_paths "/usr/include")
    list(APPEND _include_paths "/usr/src")
```

I'm not sure what the desired outcome was, but `/usr/include/{gmock,gtest}` seems like it should be included, but the [package-specific dir here](https://github.com/ros/catkin/blob/kinetic-devel/cmake/test/gtest.cmake#L258) is only appended to the last path.

Doing some output testing by modifying the [code here](https://github.com/ros/catkin/blob/kinetic-devel/cmake/test/gtest.cmake#L303):
```
    set(_include_paths "/usr/include" "/usr/src")
    set(_gmock_include_paths "${_include_paths}/gmock")
    message("### _gmock_include_paths: ${_gmock_include_paths}")
    ### _gmock_include_paths: /usr/include;/usr/src/gmock
```

Now, when I fix that to point at `/usr/include/gmock`, I get another error:
```
CMake Error at /home/jwhendy/catkin_ws/src/catkin/cmake/test/gtest.cmake:344 (add_subdirectory):
  The source directory

    /usr/src

  does not contain a CMakeLists.txt file.
Call Stack (most recent call first):
  cmake/all.cmake:164 (include)
  CMakeLists.txt:8 (include)


CMake Error at /home/jwhendy/catkin_ws/src/catkin/cmake/test/gtest.cmake:346 (set_target_properties):
  set_target_properties Can not find target to add properties to: gtest
Call Stack (most recent call first):
  cmake/all.cmake:164 (include)
  CMakeLists.txt:8 (include)


CMake Error at /home/jwhendy/catkin_ws/src/catkin/cmake/test/gtest.cmake:349 (set_target_properties):
  set_target_properties Can not find target to add properties to: gmock
Call Stack (most recent call first):
  cmake/all.cmake:164 (include)
  CMakeLists.txt:8 (include)


make: *** [Makefile:390: cmake_check_build_system] Error 1
```

Starting with the first, it's [this line](https://github.com/ros/catkin/blob/kinetic-devel/cmake/test/gtest.cmake#L341). Printing out the directories:
```
    message("### ${base_dir} ${gtest_lib_dir}")
    ### /usr/src /home/jwhendy/catkin_ws/build/catkin/gtest
```

I would not expect `CMakeLists.txt` in `/usr/src` either. If I manually change it to `/usr/src/gtest`, it goes away. Then I get an error at [this line](https://github.com/ros/catkin/blob/kinetic-devel/cmake/test/gtest.cmake#L346).
```
CMake Error at /home/jwhendy/catkin_ws/src/catkin/cmake/test/gtest.cmake:349 (set_target_properties):
  set_target_properties Can not find target to add properties to: gmock
```
Adding some messages yields:
```
    message("### ${gtest_libs} ${gtest_main_libs}")
    message("### ${gmock_libs} ${gmock_main_libs}")
    ### gtest gtest_main
    ### gmock gmock_main
```

So, we're back to the filesystem woes:
```
# all good for gtest, hence no error
$ ls /usr/src/gtest/src/
gtest-all.cc  gtest.cc  gtest-death-test.cc  gtest-filepath.cc  gtest-internal-inl.h  gtest_main.cc  gtest-port.cc  gtest-printers.cc  gtest-test-part.cc  gtest-typed-test.cc

# booo
$ ls /usr/src/gmock/src
ls: cannot access '/usr/src/gmock/src': No such file or directory
```

So, I think one could make the case that arch should align these two packages to have the same `src` location.

I still think something is off, though. If I change [this line](https://github.com/ros/catkin/blob/kinetic-devel/cmake/test/gtest.cmake#L259) to this:
```
set(_gmock_source_paths "${src_path}/gmock")
```
That error still exists, even though that's where the files are. I then copied `*.cc` to `/usr/src/gmock/src` but am still getting the same error:
```
### gmock gmock_main
CMake Error at /home/jwhendy/catkin_ws/src/catkin/cmake/test/gtest.cmake:348 (set_target_properties):
  set_target_properties Can not find target to add properties to: gmock
```

That's where I'll leave it for now. I feel extremely inefficient in tracking this down and there appears to be at least a couple independent sets of paths maintained depending on what goes hunting for `gtest`/`gmock`. This might make sense to someone, but I'm pretty confused on what ultimately needs to be found, which paths match up to what and for what purpose, and what's now breaking. This isn't the behavior I got yesterday and now I don't seem to be able to replicate my eventual success! Out of time for the day...

@dirk-thomas follow-up on the src warnings. As mentioned above, this line in [the default gmock `CMakeLists.txt`](https://github.com/google/googletest/blob/master/googlemock/CMakeLists.txt) does not work:

```
# A directory to find Google Test sources.
if (EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/gtest/CMakeLists.txt")
  set(gtest_dir gtest)
else()
  set(gtest_dir ../googletest)
endif()
```

With this default, I get:
```
Starting  >>> catkin                                                                                                                                                                     
_________________________________________________________________________________________________________________________________________________________________________________________
Errors     << catkin:check /home/jwhendy/catkin_ws/logs/catkin/build.check.002.log                                                                                                       
CMake Warning at /usr/src/gmock/CMakeLists.txt:37 (project):
  VERSION keyword not followed by a value or was followed by a value that
  expanded to nothing.


CMake Error at /usr/src/gmock/CMakeLists.txt:47 (add_subdirectory):
  add_subdirectory given source "../googletest" which is not an existing
  directory.


CMake Error at /usr/src/gmock/CMakeLists.txt:59 (config_compiler_and_linker):
  Unknown CMake command "config_compiler_and_linker".


make: *** [Makefile:446: cmake_check_build_system] Error 1
```

I didn't know what to make this, so I guessed `../gtest` instead. That does build, but leads to a warning for every package like this:
```
Starting  >>> catkin                                                                                                                                                                     
_________________________________________________________________________________________________________________________________________________________________________________________
Warnings   << catkin:check /home/jwhendy/catkin_ws/logs/catkin/build.check.003.log                                                                                                       
CMake Warning at /usr/src/gmock/CMakeLists.txt:37 (project):
  VERSION keyword not followed by a value or was followed by a value that
  expanded to nothing.


CMake Warning (dev) at /usr/src/gmock/CMakeLists.txt:47 (add_subdirectory):
  Policy CMP0013 is not set: Duplicate binary directories are not allowed.
  Run "cmake --help-policy CMP0013" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  The binary directory

    /home/jwhendy/catkin_ws/build/catkin/gtest

  is already used to build a source directory.  This command uses it to build
  source directory

    /usr/src/gtest

  which can generate conflicting build files.  CMake does not support this
  use case but it used to work accidentally and is being allowed for
  compatibility.
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning at /usr/src/gtest/CMakeLists.txt:47 (project):
  VERSION keyword not followed by a value or was followed by a value that
  expanded to nothing.
```

I'm guessing the proposal would be to take this up with googletest and arch?

**Edit:** for clarification, arch doesn't ship with a `/usr/src/gmock/CMakeLists.txt` for whatever reason. So, it will break using the default package, then also break of you google around for what to do about this missing file (i.e. download it), and give warnings when I do the above.

--- a/cmake/test/gtest.cmake
+++ b/cmake/test/gtest.cmake
@@ -212,8 +212,9 @@ endfunction()
 # Google recommends distributing GTest and GMock as source only, to be built with the same
 # flags as that which is being tested.
 #
-# :param[in] include_path: Path to search for Google Test includes
-# :param[in] src_path: Path to search for Google Test sources
+# :param[in] gtest_path: Base path to search for gtest sources and includes, eg /usr
+# :param[in] googletest_path: Base path to search for googletest-packaged gtest/gmock,
+#                             eg, /usr/src/googletest
 # :param[out] gtest_found: Whether or not GTest was found in the paths provided
 # :param[out] gtest_include_dir: The include path to access GTest's headers
 # :param[out] gtest_lib_dir: The library path to access GTest's libraries
@@ -226,16 +227,19 @@ endfunction()
 # :param[out] gmock_main_libs: GMock's main libraries
 # :param[out] base_dir: The base directory containing Google Test and/or GMock CMakeLists.txt
 #
-function(catkin_find_google_test_source include_path src_path
+function(catkin_find_google_test_source gtest_path googletest_path
   gtest_found gtest_include_dir gtest_lib_dir gtest_libs gtest_main_libs
   gmock_found gmock_include_dir gmock_lib_dir gmock_libs gmock_main_libs
   base_dir
 )
-  #Find GTest
-  set(_gtest_include_paths "${include_path}/gtest")
-  set(_gtest_source_paths "${src_path}/gtest/src")
-  list(APPEND _gtest_include_paths "${include_path}/googletest/googletest/include/gtest")
-  list(APPEND _gtest_source_paths "${src_path}/googletest/googletest/src")
+  # Path to gtest from the libgtest-dev Debian package.
+  set(_gtest_include_paths "${gtest_path}/include/gtest")
+  set(_gtest_source_paths "${gtest_path}/src/gtest/src")
+
+  # Path to gtest from the googletest Debian package.
+  list(APPEND _gtest_include_paths "${googletest_path}/googletest/include/gtest")
+  list(APPEND _gtest_source_paths "${googletest_path}/googletest/googletest/src")
+
   if(CATKIN_TOPLEVEL)
     # Ensure current workspace is searched before system path
     list(INSERT _gtest_include_paths 0 "${CMAKE_SOURCE_DIR}/googletest/googletest/include/gtest")
@@ -254,11 +258,14 @@ function(catkin_find_google_test_source include_path src_path
   set(${gtest_main_libs} ${_gtest_main_libs} PARENT_SCOPE)
   set(${base_dir} ${_gtest_base_dir} PARENT_SCOPE)
 
-  #Find GMock
-  set(_gmock_include_paths "${include_path}/gmock")
-  set(_gmock_source_paths "${src_path}/gmock/src")
-  list(APPEND _gmock_include_paths "${include_path}/googletest/googlemock/include/gmock")
-  list(APPEND _gmock_source_paths "${src_path}/googletest/googlemock/src")
+  # Path to gmock from the google-mock Debian package before v1.8.0.
+  set(_gmock_include_paths "${gtest_path}/include/gmock")
+  set(_gmock_source_paths "${gtest_path}/src/gmock/src")
+
+  # Path to gmock from the googletest Debian package.
+  list(APPEND _gmock_include_paths "${googletest_path}/googlemock/include/gmock")
+  list(APPEND _gmock_source_paths "${googletest_path}/googlemock/src")
+
   if(CATKIN_TOPLEVEL)
     # Ensure current workspace is searched before system path
     list(INSERT _gmock_include_paths 0 "${CMAKE_SOURCE_DIR}/googletest/googlemock/include/gmock")
@@ -281,9 +288,9 @@ function(catkin_find_google_test_source include_path src_path
     set(${base_dir} ${_gmock_base_dir} PARENT_SCOPE)
   endif()
 
-  #In googletest 1.8, gmock and gtest where merged inside googletest
-  #In this case, including gmock builds gmock twice, so instead we need to include googletest
-  #which will include gtest and gmock once
+  # In googletest 1.8, gmock and gtest were merged inside googletest
+  # In this case, including gmock builds gmock twice, so instead we need to include googletest
+  # which will include gtest and gmock once
   if(_gtest_found)
     get_filename_component(_gtest_base_dir_realpath ${_gtest_base_dir} REALPATH)
     get_filename_component(_global_base_dir ${_gtest_base_dir_realpath} PATH)
@@ -299,11 +306,13 @@ if(NOT GMOCK_FOUND OR NOT GTEST_FOUND)
   # If we find one but not the other, see if we can get both from source
   # only add gmock/gtest directory once per workspace
   if(NOT TARGET gtest AND NOT TARGET gmock)
-    # Fall back to system-installed gmock source (e.g. Ubuntu)
-    set(_include_paths "/usr/include" "/usr/src")
-    set(_source_paths "/usr/src")
+    # Path to base of legacy libgtest-dev and google-mock packages.
+    set(_gtest_path "/usr")
 
-    catkin_find_google_test_source("${_include_paths}" "${_source_paths}" gtest_found
+    # Path to base of new googletest package, which includes both gtest and gmock.
+    set(_googletest_path "/usr/src/googletest")
+
+    catkin_find_google_test_source("${_gtest_path}" "${_googletest_path}" gtest_found
                                    gtest_include_dir gtest_lib_dir gtest_libs gtest_main_libs
                                    gmock_found gmock_include_dir gmock_lib_dir gmock_libs
                                    gmock_main_libs base_dir)

